### DISCUSSION TITLE
Why the userParameters cannot be acquired for online-scanner reconstruction?

### DISCUSSION THREAD

**Discussion by bigtoma** [gadgetron/gadgetron#1245] [Q&A] [answered]
I use the matlab interface and icegadgetron to achieve online reconstruction in siemens scanner.
But the header.userParameters.userParameterLong cannot be acquired in the real-time reconstruction. 

If the rawdata acquisition was done, I can use the icesimu in IDEA or the retrospective reconstruction of scanner to recon. At this time, the matlab can obtain all the userParameters. 
![Snipaste_2024-01-18_15-49-04](https://github.com/gadgetron/gadgetron/assets/42628712/ac8dde6e-8d39-4b19-a4bd-66347ce54a79)

Does anyone know how to solve this problem？
Thanks!

**Comment by mfcallaghan** [comment]
Hi @bigtoma, What do you mean when you say that the parameters cannot be acquired?  Can you not retrieve these from connection.header?

In the XML configuring your Gadgetron reconstruction you might connect to Matlab via:
```
 <external>
        <execute name="MatlabRecon" type="matlab"/>
            <configuration/>
 </external>
```

Then in MatlabRecon (or a sub-function that received connection.header as input) you could extract the parameters. For example to retrieve a parameter named "caipiFactor":
```
lUserParamValues = [connection.header.userParameters.userParameterLong.value];
lUserParamNames = {connection.header.userParameters.userParameterLong.name};

if strcmp(lUserParamNames, 'caipiFactor')
    myCaipiFactor = lUserParamValues(strcmp(lUserParamNames, 'caipiFactor'));
else
    myCaipiFactor = 0;
end
```
We do this and it works fine for online reconstruction on Siemens scanners.

**Reply by bigtoma** [reply]
Thanks for your answer.

Yes, the value of variable in the header.userParameters.userParameterLong mostly are null for my online reconstruction. 

 I found that MATLAB can read the original parameters of the sequence such as TE TR or acceleration factor, but not fot the parameters I added. And I use parameter_map3 to add the userParameters.  It is in the special card of the sequence UI, and its name will always be like sWipMemBlock.xxxx. What configuration file needs to be modified for the newly added parameters, such as IsmrmrdParameterMap_Siemens.xml?

How do you add the userParameters in the sequence？ Maybe there's something wrong with the way I added userParameters.

**Comment by mfcallaghan** [comment, marked as answer]
We don’t use parameter_map3 but instead use Siemens’ WIPParameterTool.  Nonetheless, with sWipMemBlock.xxx filled, we read them as follows:

In IsmrmrdParameterMap_Siemens.xml we have:
  ```
<p><s>MEAS.sWipMemBlock.alFree</s>                          <d>siemens.MEAS.sWipMemBlock.alFree</d></p>
 <p><s>MEAS.sWipMemBlock.adFree</s>                          <d>siemens.MEAS.sWipMemBlock.adFree</d></p>
```

And in IsmrmrdParameterMap_Siemens_EPI_wip_VE12.xsl we extract the values or provide defaults, e.g. extracting the CAIPI factor or number of segments:
```
                   <userParameterLong>
                      <name>CAIPI</name>
				<value>
					<xsl:choose>
		         			<xsl:when test="not(siemens/MEAS/sWipMemBlock/alFree[1])">0</xsl:when>
            				<xsl:otherwise>
                				<xsl:value-of select="siemens/MEAS/sWipMemBlock/alFree[1]"/> 
		            		</xsl:otherwise>
        				</xsl:choose>
				</value>		
                    </userParameterLong>
		    <userParameterLong>
                        <name>nSeg</name>
				<value>
					<xsl:choose>
		         			<xsl:when test="not(siemens/MEAS/sWipMemBlock/alFree[2])">1</xsl:when>
            				<xsl:otherwise>
                				<xsl:value-of select="siemens/MEAS/sWipMemBlock/alFree[2]"/> 
		            		</xsl:otherwise>
        				</xsl:choose>
				</value>		
                    </userParameterLong>
```

**Reply by bigtoma** [reply]
Thank you so much, I will try it.

**Reply by bigtoma** [reply]
Oh, I found that my IsmrmrdParameterMap_Siemens.xsl was the culprit. The IsmrmrdParameterMap_Siemens.xsl I used was the original xsl generated from the IceGadgetron, which didn't put all the alFree[] into the header.

If I use the IsmrmrdParameterMap_Siemens.xsl generated from ismrmrd, I can get all things right because it read all alFree[] by 

> 		    
                <xsl:for-each select="siemens/MEAS/sWipMemBlock/alFree">
                    
                      <xsl:variable name="CurLong" select="position()"/>
                      <userParameterLong>
                        <name>sWipMemBlock.alFree[<xsl:value-of select="$CurLong - 1" />]</name>
                        <value>
                          <xsl:value-of select="."/>
                        </value>
                      </userParameterLong>
                    
                </xsl:for-each>

Thanks again!
